### ISSUE TITLE
Resume training out of memory

### ISSUE THREAD

**Issue by shahabe** [dbolya/yolact#296]
@dbolya I trained a custom model for many iterations and want to resume the training by adding more images. When I run 
```
python train.py --config=yolact_base_config --resume=weights/yolact_base_3_345_interrupt.pth --start_iter=-1
``` 
it gives me the cuda out of memory error:
```
RuntimeError: CUDA out of memory. Tried to allocate 40.00 MiB (GPU 0; 5.80 GiB total capacity; 3.63 GiB already allocated; 16.81 MiB free; 80.22 MiB cached)
```

I trained the model using the same machine and setup.
Do you have some siggestion on this issue?
Thank you.

**Comment by dbolya** [comment]
Looks like you ran out of GPU memory. Either reduce the batch size or use a less beefy model (i.e., the resnet50 version).

You can reduce the batch size with `--batch_size` without retraining, so I would try that first.

**Comment by shahabe** [comment]
Yes, it ran out of gpu memory, but the problem is that I try to resume the training with exactly same setup. It means the training was started before and everything was alright. 

I think when it tries to load the interrupted version of the weights, uses more GPU memory. But I cannot find the reason.

**Comment by dbolya** [comment]
Well it says:
```
5.80 GiB total capacity
3.63 GiB already allocated
16.81 MiB free
```
which means that your card at the time of running the program only had 3.64 GiB available out of the 5.80 GiB total capacity. That means that other programs are using this card (e.g., your OS, YouTube, etc.). These other programs can use more or less vram depending on the specific content onscreen.

It could be that loading weights from memory uses more vram as well, but can you try starting to train from scratch and then resuming to make sure? Because it also could be that you're just using your GPU for more things in the background than when you started training initially.

**Comment by shahabe** [comment]
@dbolya Thank you for your reply.
I managed to resume the training using another GPU with more RAM.
However, I believe there is an issue with resuming the training session:

I keep monitoring the GPU memory usage using `watch -n0.1 nvidia-smi`.
GPU memory usage is 0MiB (I am not using any graphics working on aws instance).
When I start the training from scratch, the GPU memory usage is 2954MiB out of  total 11441MiB.
I stop the process and it saves `yolact_base_37_5253_interrupt.pth` file.
Now the GPU memory usage is 0MiB again.
Then I use this file to resume the training as
```
python train.py --config=yolact_base_config --resume=weights/yolact_base_37_5253_interrupt.pth --start_iter=-1
```

When the training resumes, the memory usage becomes 10244MiB. 

That is why I assume that in the resuming process, the GPU memory is loading unnecessary data.
What do you think about it?

**Comment by dbolya** [comment]
Wow that's a big difference! Though, what batch size is that? Since if you're using the default batch size of 8, I'd expect yolact_base_config to be using 10+gb of vram even when training from scratch. Are you sure you're profiling the first number correctly? 2 gb seems way too low.

**Comment by shahabe** [comment]
My images' size are 550X550 and I changed the batch size to 2
Yes, even in my local machine it consumes the same amount of vram.
I wanted to resume the training but had that issue. That is why I checked with my AWS instances which has 12GB vram.

I truly appreciate your time to check the code to see where the possible large memory assumption is happening.

**Comment by dbolya** [comment]
Wait that's 10gb with a batch size of 2? Uhh that would be a problem.

I'm definitely not getting that big of a difference on my machines, but I'll test to see if there's any difference on my side.

**Comment by shahabe** [comment]
@dbolya Another issue with resuming:
I start the resuming process by:
```
python train.py --config=yolact_base_config --resume=weights/yolact_base_319_180392_interrupt.pth --start_iter=-1
```
which is on epoche 319 and iteration 180392 (correct me if I am wrong).
But when I resume the training it starts from epoche 1279:
![Screenshot from 2020-02-05 12-06-26](https://user-images.githubusercontent.com/3667841/73802608-4146ed80-4812-11ea-84cb-e85234fd4296.png)
I have almost 1250 images all with the size 550X550, batch size = 2

**Comment by dbolya** [comment]
@shahabe That could be an artifact of changing batch size, since the epoch size is dependent on batch size. It shouldn't affect anything though, since all the stopping conditions are based on the iterations and that stays consistent.

---

Also I tested your issue like this:

I started by launching ```watch -n0.1 nvidia-smi``` in another window.

Then I set `CUDA_VISIBLE_DEVICES` to 0 (a GPU with no graphics) and confirmed that GPU 0 had 0 vram usage.

Then I ran
```python train.py --batch_size=2 --config=yolact_base_config```
and then let it train for ~100 iterations before I stopped it with ctrl+c. It reached a stable memory usage of 3619 MiB.

Finally, I waited for GPU 0 to once again have 0 vram usage and then ran
```python train.py --batch_size=2 --config=yolact_base_config --resume=interrupt --start_iter=-1```
After the memory usage stabilized, this used 3611 MiB of vram (though it's been slowly creeping up by 2 MiB incriments, so maybe it will go back to 3619).

I tried the same for `yolact_plus_base_config` and got `3965 MiB` the first run and `3857 MiB` from the resume.

---

So I can't reproduce what you got. Note that the amount of memory Pytorch uses is not the same as what the model is using, since it reserves extra memory to have on hand as cache. It also matters how frequently the python garbage collector runs, because that's when the GPU memory gets freed.

Note also that my memory usage was higher than yours even for the same batch size and model. This is probably because of differences in caching based on how much vram is available in the system or something.

Idk why you got 10 GiB that one time, but it could be Pytorch doing something weird with caching. It's a little frustrating to have such inconsistent results with memory usage, but ¯\\\_(ツ)\_/¯

**Comment by shahabe** [comment]
@dbolya Thank you so much for your information.
For vram usage I did exactly the same as you, except setting `CUDA_VISIBLE_DEVICES` to 0 (which I am not sure what it is for?) I also waited to see the 0 GPU usage, when I stop training and then resume the training. It is strange anyway.

Other possibilities can be the difference between the pytorch version or other required packages.
Would you please send me the list of the packages in your environment to compare?

______

About the wrong number of epochs, I have not changed the batch size and it is 2 for both training and resuming. Anyway, as you said, that is not a big issue. Thank you for that as well.

**Comment by dbolya** [comment]
`CUDA_VISIBLE_DEVICES` sets which GPU to use in the case you have multiple GPUs. If you only have one GPU, it shouldn't matter. So uh, this is weird.

My environment is:
```
Ubuntu 18.04
Python 3.7.3
CUDA 10.1.243
CuDNN 7603
```

Of the libraries that YOLACT uses, I have:
```
Cython==0.29.13
matplotlib==3.1.1
numpy==1.17.2
opencv-python==4.1.1.26
Pillow==6.2.0
pycocotools==2.0.0
torch==1.3.0
torchvision==0.4.1
```

Note that you might be able to squeeze out more vram by downgrading to `torch==1.0.1` and whatever torchvision package is appropriate for that version. We noticed a jump in vram usage from 1.0.1 to 1.1.0.
